sklearn/_build_utils/__init__.py

```diff
@@ -70,20 +71,25 @@ def cythonize_extensions(extension):
         os.environ.get("SKLEARN_ENABLE_DEBUG_CYTHON_DIRECTIVES", "0") != "0"
     )
 
+    compiler_directives = {
+        "language_level": 3,
+        "boundscheck": cython_enable_debug_directives,
+        "wraparound": False,
+        "initializedcheck": False,
+        "nonecheck": False,
+        "cdivision": True,
+    }
+
+    if not Cython.__version__.startswith("0."):
```

Comment:
This setting will not allow scikit-learn to build with any of Cython's alpha releases because `legacy_implicit_noexcept` has not been released in any of the alphas. I wanted to suggestion this:

```python
    if parse(Cython.__version__) >= parse("3.0.0a11"):
```

But the dev branch of Cython is still "3.0.0a11": https://github.com/cython/cython/blob/master/Cython/Shadow.py#L4, which is the same the most recent [PyPi alpha release](https://pypi.org/project/Cython/#history).

I am okay with the compromise of only allowing scikit-learn to build with Cython 3 on the dev branch.

Reply:
FYI, I have opened https://github.com/cython/cython/issues/5204 to ask for a support of Development release segment in public version identifiers [as specified by PEP 440](https://peps.python.org/pep-0440/#public-version-identifiers).

Reply:
Now that https://github.com/cython/cython/issues/5204 has been addressed, we should be able to use this:
```suggestion
    if not Cython.__version__.startswith("3.0.0a12.dev0"):
```

Reply:
Your PR was likely merged too quickly and broke installing cython from the master branch, see https://github.com/cython/cython/issues/5204#issuecomment-1377014880

Reply:
https://github.com/cython/cython/pull/5208 should fix that.